File: machine.py
#! /usr/bin/env python3

import struct
import itertools
from array import array as new_array

memory = {}
next_index = 0
abandoned_indexes = set()
reg = [0] * 8
finger = 0

def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('program')
    parser.add_argument('-c', '--cycles', type=int,
        help='number of cycles to execute')
    args = parser.parse_args()

    with open(args.program, 'rb') as fp:
        memory[0] = load(fp)

    cycle(args.cycles)

def load(fp):
    array = new_array('I')
    buf = fp.read(4)
    while buf:
        platter = struct.unpack('>I', buf)[0]
        array.append(platter)
        buf = fp.read(4)
    return array

def cycle(limit=None):
    global finger
    i = 0
    code = None
    args = None

    try:
        for i in itertools.count():
            platter = memory[0][finger]
            code, args = unpack_op(platter)

            op = operators[code]
            op(*args)

            if limit is not None and i == limit:
                state('LIMIT', i, code, args)
                break

            finger += 1
    except Exception:
        state('FAIL', i, code, args)
        raise
    except:
        state('\nEXIT', i, code, args)
        raise

def unpack_op(platter):
    code = platter >> 28

    if code == 13:
        a = (platter >> 25) & 7
        value = platter & 0x1ffffff
        args = (a, value)
    else:
        a = (platter >> 6) & 7
        b = (platter >> 3) & 7
        c = platter & 7
        args = (a, b, c)

    return code, args

def state(msg, cycle=0, code=None, args=None):
    print('{} at {} (cycle {})'.format(msg, finger, cycle))
    if code is not None:
        name = operators[code].__name__
        print('{}|{}{}'.format(name, code, args))
    print('reg', reg)
    print('memory', {i: len(a) for i, a in itertools.islice(memory.items(), 10)})

##
## operators
##

def cmv(a, b, c):
    """
    #0. Conditional Move.

    The register A receives the value in register B, unless the register C
    contains 0.
    """
    if reg[c] != 0:
        reg[a] = reg[b]

def aix(a, b, c):
    """
    #1. Array Index.

    The register A receives the value stored at offset in register C in the
    array identified by B.
    """
    array = memory[reg[b]]
    reg[a] = array[reg[c]]

def aam(a, b, c):
    """
    #2. Array Amendment.

    The array identified by A is amended at the offset in register B to store
    the value in register C.
    """
    array = memory[reg[a]]
    array[reg[b]] = reg[c]

def add(a, b, c):
    """
    #3. Addition.

    The register A receives the value in register B plus the value in register
    C, modulo 2^32.
    """
    reg[a] = (reg[b] + reg[c]) % (2**32)

def mul(a, b, c):
    """
    #4. Multiplication.

    The register A receives the value in register B times the value in register
    C, modulo 2^32.
    """
    reg[a] = (reg[b] * reg[c]) % (2**32)

def div(a, b, c):
    """
    #5. Division.

    The register A receives the value in register B divided by the value in
    register C, if any, where each quantity is treated treated as an unsigned
    32 bit number.
    """
    reg[a] = reg[b] // reg[c]

def nad(a, b, c):
    """
    #6. Not-And.

    Each bit in the register A receives the 1 bit if either register B or
    register C has a 0 bit in that position.  Otherwise the bit in register A
    receives the 0 bit.
    """
    reg[a] = (reg[b] & reg[c]) ^ ((2**32) - 1)

def hlt(_a, _b, _c):
    """
    #7. Halt.

    The universal machine stops computation.
    """
    raise SystemExit

def alc(_a, b, c):
    """
    #8. Allocation.

    A new array is created with a capacity of platters commensurate to the
    value in the register C. This new array is initialized entirely with
    platters holding the value 0. A bit pattern not consisting of exclusively
    the 0 bit, and that identifies no other active allocated array, is placed
    in the B register.
    """
    global next_index
    try:
        index = abandoned_indexes.pop()
    except KeyError:
        next_index += 1
        index = next_index
    memory[index] = new_array('I', [0] * reg[c])
    reg[b] = index

def abd(_a, _b, c):
    """
    #9. Abandonment.

    The array identified by the register C is abandoned. Future allocations
    may then reuse that identifier.
    """
    index = reg[c]
    del memory[index][:]
    abandoned_indexes.add(index)

def out(_a, _b, c):
    """
    #10. Output.

    The value in the register C is displayed on the console immediately. Only
    values between and including 0 and 255 are allowed.
    """
    print(chr(reg[c]), end='')

def inp():
    """
    #11. Input.

    The universal machine waits for input on the console. When input arrives,
    the register C is loaded with the input, which must be between and
    including 0 and 255. If the end of input has been signaled, then the
    register C is endowed with a uniform value pattern where every place is
    pregnant with the 1 bit.
    """
    pass

def lod(_a, b, c):
    """
    #12. Load Program.

    The array identified by the B register is duplicated and the duplicate
    shall replace the '0' array, regardless of size. The execution finger is
    placed to indicate the platter of this array that is described by the
    offset given in C, where the value 0 denotes the first platter, 1 the
    second, et cetera.

    The '0' array shall be the most sublime choice for loading, and shall be
    handled with the utmost velocity.
    """
    global finger

    index = reg[b]
    if index != 0:
        memory[0] = memory[index][:]

    finger = reg[c] - 1

def ort(a, value):
    """
    #13. Orthography.

    The value indicated is loaded into the register A forthwith.
    """
    reg[a] = value

operators = [
    cmv, aix, aam, add, mul, div, nad, hlt, alc, abd, out, inp, lod, ort]

##
## entry
##

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        pass

File: machine.fs
open System
open System.IO

type Op =
    | cmv=0 | aix=1 | aam=2 | add=3 | mul=4 | dvi=5 | nad=6 | hlt=7 | alc=8
    | abd=9 | out=10 | inp=11 | lod=12 | ort=13

let mutable memory = new ResizeArray<uint32 array>()
let mutable abandoned_indexes = new ResizeArray<uint32>()
let mutable reg: uint32 array = Array.zeroCreate 8

let load filename =
    let mutable program: uint32 array =
        Array.zeroCreate (int (FileInfo(filename).Length / 4L))
    let fp = File.Open(filename, FileMode.Open, FileAccess.Read, FileShare.Read)
    let reader = new BinaryReader(fp)
    let mutable index = 0
    try
        while true do
            let platter = reader.ReadInt32() |> System.Net.IPAddress.HostToNetworkOrder
            program.[index] <- uint32 platter
            index <- index + 1
    with
    | :? System.IO.EndOfStreamException -> ()

    memory.Add program

let state code a b c =
    printfn "%u(%u, %u, %u)" code a b c
    printfn "reg: %A" reg
    printfn "mem: %A" [for x in memory do yield x.Length]
    printfn "abd: %A" abandoned_indexes

let run () =
    let mutable finger = 0ul
    let mutable cycle = 0ul
    let mutable stop = false

    let mutable platter = 0ul
    let mutable code = 0uy
    let mutable a = 0uy
    let mutable b = 0uy
    let mutable c = 0uy
    let mutable value = 0ul

    while not stop && (int finger) < memory.[0].Length do
        // printfn "cycle: %u" cycle
        platter <- memory.[0].[int finger]
        code <- uint8 (platter >>> 28)

        match enum<Op>(int32 code) with
        | Op.ort ->
            a <- (uint8 (platter >>> 25)) &&& 7uy
            value <- platter &&& 0x1fffffful
            // printfn "%u(%u, %u)" code a value
            reg.[int a] <- value
        | _ ->
            a <- (uint8 (platter >>> 6)) &&& 7uy
            b <- (uint8 (platter >>> 3)) &&& 7uy
            c <- (uint8 platter) &&& 7uy

            match enum<Op>(int32 code) with
            | Op.cmv -> if reg.[int c] <> 0ul then reg.[int a] <- reg.[int b]
            | Op.aix -> reg.[int a] <- memory.[int reg.[int b]].[int reg.[int c]]
            | Op.aam -> memory.[int reg.[int a]].[int reg.[int b]] <- reg.[int c]
            | Op.add -> reg.[int a] <- reg.[int b] + reg.[int c]
            | Op.mul -> reg.[int a] <- reg.[int b] * reg.[int c]
            | Op.dvi -> reg.[int a] <- reg.[int b] / reg.[int c]
            | Op.nad -> reg.[int a] <- ~~~(reg.[int b] &&& reg.[int c])
            | Op.hlt -> stop <- true
            | Op.alc ->
                let new_array = Array.zeroCreate (int reg.[int c])
                let index =
                    if abandoned_indexes.Count > 0 then
                        let i = abandoned_indexes.[0]
                        abandoned_indexes.RemoveAt(0)
                        memory.[int i] <- new_array
                        i
                    else
                        memory.Add new_array
                        uint32 memory.Count - 1ul
                reg.[int b] <- index
            | Op.abd ->
                memory.[int reg.[int c]] <- Array.empty
                abandoned_indexes.Add reg.[int c]
            | Op.out ->
                printf "%c" (char reg.[int c])
                stdout.Flush()
            | Op.inp -> reg.[int c] <- uint32 (Console.ReadKey()).KeyChar
            | Op.lod ->
                if reg.[int b] <> 0ul then
                    memory.[0] <- Array.copy memory.[int reg.[int b]]
                finger <- reg.[int c] - 1ul
            | _ ->
                printfn "unknown code: %u" code
                stop <- true

        finger <- finger + 1ul
        cycle <- cycle + 1ul


// allow ctrl-c to kill
System.Console.CancelKeyPress.Add(fun _ -> ())

load @"scrolls/codex.umz"
run()
